README
This is Perl module Data::Wheren.

INSTALLATION

Type the following command:

    $ curl -L http://cpanmin.us | perl - Data::Wheren

Or install cpanm and then run the following command to install
Data::Wheren:

    $ cpanm Data::Wheren

If you get an archive of this distribution, unpack it and build it
as per the usual:

    $ tar xzf Data-Wheren-$version.tar.gz
    $ cd Data-Wheren-$version
    $ perl Makefile.PL
    $ make && make test

Then install it:

    $ make install

DOCUMENTATION

Data::Wheren documentation is available as in POD. So you can do:

    $ perldoc Data::Wheren

to read the documentation online with your favorite pager.

LICENSE AND COPYRIGHT

Copyright (c) 2011, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Data::Wheren

NEXT <<SET RELEASE DATE HERE>>
    - original version; created by dim
      at Sat Oct 15 21:49:54 2011.


Makefile.PL
#!perl
use strict;
use warnings;
BEGIN {
    unshift @INC, 'inc';

    # author requires, or bundled modules
    my @devmods = qw(
        inc::Module::Install             1.00
        Module::Install::AuthorTests     0.002
        Module::Install::Repository      0.06
        Test::Requires                   0.06
    );
    my @not_available;
    while(my($mod, $ver) = splice @devmods, 0, 2) {
        eval qq{use $mod $ver (); 1} or push @not_available, $mod;
    }
    if(@not_available) {
        print qq{# The following modules are not available.\n};
        print qq{# `perl $0 | cpanm` will install them:\n};
        print $_, "\n" for @not_available;
        exit 1;
     }
}
use inc::Module::Install;

all_from 'lib/Data/Wheren.pm';


test_requires 'Test::More'     => '0.88';
test_requires 'Test::Requires' => '0.06';

auto_set_repository;

tests_recursive;
author_tests 'xt';


clean_files qw(
    Data-Wheren-*
    *.stackdump
    cover_db *.gcov *.gcda *.gcno
    nytprof
    *.out
);

WriteAll(check_nmake => 0);


lib/Data/Wheren.pm
package Data::Wheren;
use 5.008_001;
use strict;
use warnings;

our $VERSION = '0.01';

use Carp;

sub new {
    my ($class, %args) = @_;

    croak "should override" if $class eq "Data::Wheren";

    $args{min} ||= 0;
    $args{max} ||= 2147451247; # 2038-01-19 03:14:7
    $args{dec} = +{ map { $args{enc}->[$_] => $_ } 0 .. scalar @{$args{enc}} - 1};

    my $self = bless \%args, $class;

    $self->{neighbor} = $self->_gen_neihbor();
    $self->{border} = $self->_gen_border();

    $self;
}

sub _gen_int {
    my $self = shift;
    [ [ -90, 90 ], [ -180, 180 ], [ $self->{min}, $self->{max} ] ];
}

#my ($self, $lat, $lon, $timestamp, $level) = @_;
sub encode {
    croak "should be override encode method";
}

sub decode {
    my ($self, $str) = @_;

    my $int = $self->decode_to_interval($str);

    map { ( $_->[0] + $_->[1] ) / 2 } @$int;
}

sub decode_to_interval {
    croak "should be override decode_to_interval method";
}

sub adjacent {
    my ($self, $str, $direction) = @_;

    $direction = lc $direction;
    my $base = substr($str, 0, - $self->{enc_world_length});
    my $last_char = substr($str, -$self->{enc_world_length}, $self->{enc_world_length});

    croak "not exist lat char str:$str" unless defined $self->{dec}->{$last_char};
    my $index = $self->{dec}->{$last_char};

    if ( defined $self->{border}->{$direction}{$index} ) {
        $base = $self->adjacent($base, $direction);
    }

    $base . $self->{enc}->[$self->{neighbor}->{$direction}[$index]];
}

sub _search_data {
    my ($self, $target, $func) = @_;

    my $floor_index = 0;
    for my $floor ( @{$self->{data}} ) {

        my $row_index = 0;
        for my $row ( @$floor ) {

            my $v_index = 0;
            for my $v ( @$row ) {
                if ( $v == $target ) {
                    return $func->($floor_index, $row_index, $v_index);
                }
                $v_index++;
            }
            $row_index++;
        }
        $floor_index++;
    }
    die "error in $target";
}

sub _gen_neihbor {
    my ($self) = @_;

    my $neighbor = {};

    for my $i ( 0 .. scalar @{$self->{enc}} - 1 ) {

        #right
        my $right_list = $neighbor->{right} ||= [];
        push @$right_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $self->{data}->[$floor_index][$row_index][ ($v_index + 1) % 4];
            });

        # left 
        my $left_list = $neighbor->{left} ||= [];
        push @$left_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $v_index - 1;
                $n = 3 if $n < 0;
                return $self->{data}->[$floor_index][$row_index][ $n ];
            });

        #back
        my $back_list = $neighbor->{back} ||= [];
        push @$back_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $self->{data}->[$floor_index][ ($row_index + 1) % 4 ][$v_index];
            });

        #front
        my $front_list = $neighbor->{front} ||= [];
        push @$front_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $row_index - 1;
                $n = 3 if $n < 0;
                return $self->{data}->[$floor_index][$n][$v_index];
            });

        #top
        my $top_list = $neighbor->{top} ||= [];
        push @$top_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $self->{data}->[ ($floor_index + 1) % scalar @{$self->{data}} - 1][$row_index][$v_index];
            });

        #bottom
        my $bottom_list = $neighbor->{bottom} ||= [];
        push @$bottom_list, $self->_search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $floor_index - 1;
                $n = scalar @{$self->{data}} - 1 if $n < 0;
                return $self->{data}->[$n][$row_index][$v_index];
            });
    }

    $neighbor;
}

sub _gen_border {
    my ($self) = @_;

    my $border = { right => {}, left => {}, top => {}, bottom => {}, back => {}, front => {} };

    for my $i ( 0 .. 3 ) {
        $border->{bottom}{$_} = 1 for @{$self->{data}->[0][$i]};
        $border->{top}{$_} = 1 for @{$self->{data}->[3][$i]};
    }

    for my $floor ( @{$self->{data}} ) {

        $border->{front}{$_} = 1 for @{$floor->[0]};
        $border->{back}{$_} = 1 for @{$floor->[3]};

        for my $row ( @$floor ) {
            $border->{left}{$row->[0]} = 1;
            $border->{right}{$row->[3]} = 1;
        }
    }

    $border;
}

1;
__END__

=head1 NAME

Data::Wheren - Perl extention to do something

=head1 VERSION

This document describes Data::Wheren version 0.01.

=head1 SYNOPSIS

    use Data::Wheren;

=head1 DESCRIPTION

# TODO

=head1 INTERFACE

=head2 Functions

=head3 C<< hello() >>

# TODO

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

<<YOUR NAME HERE>> E<lt><<YOUR EMAIL ADDRESS HERE>>E<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Data/Wheren/6Bit.pm
package Data::Wheren::6Bit;
use 5.008_001;
use strict;
use warnings;

our $VERSION = '0.01';

use parent 'Data::Wheren';
use Carp;

sub new {
    my ($class, %args) = @_;

    $args{enc} ||= [qw(
        A B C D E F G H I J K L M N O P Q R S T U V W X Y Z a b c d e f g h i j k l m n o p q r s t u v w x y z 0 1 2 3 4 5 6 7 8 9 ! -
        )];
    $args{enc_world_length} ||= 1;

    $args{data} ||= [
        [
            [0,2,16,18],
            [4,6,20,22],
            [32,34,48,50],
            [36,38,52,54],
        ],
        [
            [1,3,17,19],
            [5,7,21,23],
            [33,35,49,51],
            [37,39,53,55],
        ],
        [
            [0,2,16,18],
            [4,6,20,22],
            [32,34,48,50],
            [36,38,52,54],
        ],
        [
            [1,3,17,19],
            [5,7,21,23],
            [33,35,49,51],
            [37,39,53,55],
        ],
    ];
    
    $class->SUPER::new(%args);
}

#my ($self, $lat, $lon, $timestamp, $level) = @_;
sub encode {
    my ($self, @pos) = @_;
    my $level = pop @pos;

    my $int = $self->_gen_int;
    my @enc = ();

    for my $i ( 1 .. $level ) {
        my $bits = 0;
        for my $j ( 0 .. 5 ) {
            my $flag = $j % 3;

            my $mid = ($int->[$flag][0] + $int->[$flag][1]) / 2;
            my $bit = $pos[$flag] >= $mid ? 1 : 0;

            if ( $bit ) {
                $int->[$flag][0] = $mid;
            }
            else {
                $int->[$flag][1] = $mid;
            }
            $bits = ( ( $bits << 1 ) | $bit );
        }
        push @enc, $self->{enc}[$bits];
    }

    join "", @enc;
}

sub decode_to_interval {
    my ($self, $str) = @_;

    my $int = $self->_gen_int;

    for my $ch ( split //, $str ) {
        if ( defined ( my $bits = $self->{dec}{$ch} ) ) {
            for my $j ( 0 .. 5 ) {
                my $flag = $j % 3;
                $int->[$flag][ ( $bits & 32 ) >> 5 == 0 ? 1 : 0 ] = ( $int->[$flag][0] + $int->[$flag][1] ) / 2;
                $bits <<= 1;
            }
        }
        else {
            croak "Bad character '$ch' in hash '$str'";
        }
    }

    $int;
}

1;
__END__

=head1 NAME

Data::Wheren::6Bit - Perl extention to do something

=head1 VERSION

This document describes Data::Wheren::6Bit version 0.01.

=head1 SYNOPSIS

    use Data::Wheren::6Bit;

=head1 DESCRIPTION

# TODO

=head1 INTERFACE

=head2 Functions

=head3 C<< hello() >>

# TODO

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

<<YOUR NAME HERE>> E<lt><<YOUR EMAIL ADDRESS HERE>>E<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Data/SpatioTemporalHash.pm
package Data::SpatioTemporalHash;
use 5.008_001;
use strict;
use warnings;

our $VERSION = '0.01';

use Carp;

our @ENC = qw(
  A B C D E F G H I J K L M N O P Q R S T U V W X Y Z a b c d e f g h i j k l m n o p q r s t u v w x y z 0 1 2 3 4 5 6 7 8 9 ! -
  );
our $ENC_WORD_LENGTH = 1;

our %DEC = map { $ENC[$_] => $_ } 0 .. $#ENC;

our $DATA = [
    [
        [0, 4, 32, 36],
        [2, 6, 34, 38],
        [16, 20, 48, 52],
        [18, 22, 50, 54]
    ],
    [
        [1, 5, 33, 37],
        [3, 7, 35, 39],
        [17, 21, 49, 53],
        [19, 23, 51, 55]
    ],
    [
        [8, 12, 40, 44],
        [10, 14, 42, 46],
        [24, 28, 56, 60],
        [26, 30, 58, 62]
    ],
    [
        [9, 13, 41, 45],
        [11, 15, 43, 47],
        [25, 29, 57, 61],
        [27, 31, 59, 63]
    ]
];

our $NEIGHBOR = _gen_neihbor();
our $BORDER = _gen_border();

sub new {
    my ($class, %args) = @_;

    $args{min_timestamp} ||= 0;
    $args{max_timestamp} ||= 2147451247; # 2038-01-19 03:14:7

    bless \%args, $class;
}

#my ($self, $lat, $lon, $timestamp, $level) = @_;
sub encode {
    my ($self, @pos) = @_;
    my $level = pop @pos;

    my $int = [ [ -90, 90 ], [ -180, 180 ], [ $self->{min_timestamp}, $self->{max_timestamp} ] ];
    my @enc = ();

    for my $i ( 1 .. $level ) {
        my $bits = 0;
        for my $j ( 0 .. 5 ) {
            my $flag = $j % 3;

            my $mid = ($int->[$flag][0] + $int->[$flag][1]) / 2;
            my $bit = $pos[$flag] >= $mid ? 1 : 0;

            if ( $bit ) {
                $int->[$flag][0] = $mid;
            }
            else {
                $int->[$flag][1] = $mid;
            }
            $bits = ( ( $bits << 1 ) | $bit );
        }
        push @enc, $ENC[$bits];
    }

    join "", @enc;
}

sub decode {
    my ($self, $str) = @_;

    my $int = $self->decode_to_interval($str);

    map { ( $_->[0] + $_->[1] ) / 2 } @$int;
}

sub decode_to_interval {
    my ($self, $str) = @_;

    my $int = [ [ -90, 90 ], [ -180, 180 ], [ $self->{min_timestamp}, $self->{max_timestamp} ] ];

    for my $ch ( $str =~ /.{$ENC_WORD_LENGTH}/g ) {
        if ( defined ( my $bits = $DEC{$ch} ) ) {
            for my $j ( 0 .. 5 ) {
                my $flag = $j % 3;
                $int->[$flag][ ( $bits & 32 ) >> 5 == 0 ? 1 : 0 ] = ( $int->[$flag][0] + $int->[$flag][1] ) / 2;
                $bits <<= 1;
            }
        }
        else {
            croak "Bad character '$ch' in hash '$str'";
        }
    }

    $int;
}

sub adjacent {
    my ($self, $str, $direction) = @_;

    $direction = lc $direction;
    my $base = substr($str, 0, -1);
    my $last_char = substr($str, -1, 1);

    my $index = $DEC{$last_char} or croak "";

    if ( defined $BORDER->{$direction}{$index} ) {
        $base = $self->adjacent($base, $direction);
    }

    #warn $NEIGHBOR->{$direction}->[$index];
    $base . $ENC[$NEIGHBOR->{$direction}[$index]];
}

sub _search_data {
    my ($target, $func) = @_;

    my $floor_index = 0;
    for my $floor ( @$DATA ) {

        my $row_index = 0;
        for my $row ( @$floor ) {

            my $v_index = 0;
            for my $v ( @$row ) {
                if ( $v == $target ) {
                    return $func->($floor_index, $row_index, $v_index);
                }
                $v_index++;
            }
            $row_index++;
        }
        $floor_index++;
    }
    die "error in $target";
}

sub _gen_neihbor {

    my $neighbor = {};

    for my $i ( 0 .. 63 ) {

        #right
        my $right_list = $neighbor->{right} ||= [];
        push @$right_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $DATA->[$floor_index][$row_index][ ($v_index + 1) % 4];
            });

        # left 
        my $left_list = $neighbor->{left} ||= [];
        push @$left_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $v_index - 1;
                $n = 3 if $n < 0;
                return $DATA->[$floor_index][$row_index][ $n ];
            });

        #back
        my $back_list = $neighbor->{back} ||= [];
        push @$back_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $DATA->[$floor_index][ ($row_index + 1) % 4 ][$v_index];
            });

        #front
        my $front_list = $neighbor->{front} ||= [];
        push @$front_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $row_index - 1;
                $n = 3 if $n < 0;
                return $DATA->[$floor_index][$n][$v_index];
            });

        #top
        my $top_list = $neighbor->{top} ||= [];
        push @$top_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                return $DATA->[ ($floor_index + 1) % 4][$row_index][$v_index];
            });

        #bottom
        my $bottom_list = $neighbor->{bottom} ||= [];
        push @$bottom_list, _search_data($i, sub {
                my ($floor_index, $row_index, $v_index) = @_;
                my $n = $floor_index - 1;
                $n = 3 if $n < 0;
                return $DATA->[$n][$row_index][$v_index];
            });
    }

    $neighbor;
}

sub _gen_border {
    my ($self) = @_;

    my $border = { right => {}, left => {}, top => {}, bottom => {}, back => {}, front => {} };

    for my $i ( 0 .. 3 ) {
        $border->{bottom}{$_} = 1 for @{$DATA->[0][$i]};
        $border->{top}{$_} = 1 for @{$DATA->[3][$i]};
    }

    for my $floor ( @$DATA ) {

        $border->{front}{$_} = 1 for @{$floor->[0]};
        $border->{back}{$_} = 1 for @{$floor->[3]};

        for my $row ( @$floor ) {
            $border->{left}{$row->[0]} = 1;
            $border->{right}{$row->[3]} = 1;
        }
    }

    $border;
}

1;
__END__

=head1 NAME

Data::SpatioTemporalHash - Perl extention to do something

=head1 VERSION

This document describes Data::SpatioTemporalHash version 0.01.

=head1 SYNOPSIS

    use Data::SpatioTemporalHash;

=head1 DESCRIPTION

# TODO

=head1 INTERFACE

=head2 Functions

=head3 C<< hello() >>

# TODO

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

<<YOUR NAME HERE>> E<lt><<YOUR EMAIL ADDRESS HERE>>E<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Data/Wheren/7Bit.pm
package Data::Wheren::7Bit;
use 5.008_001;
use strict;
use warnings;

our $VERSION = '0.01';

use parent 'Data::Wheren';
use Carp;

sub new {
    my ($class, %args) = @_;

    $args{enc} ||= [ qw(
        00 01 02 03 10 11 12 13 20 21 22 23 30 31 32 33 40 41 42 43 50 51 52 53 60 61 62 63 70 71 72 73 80 81 82 83 90 91 92 93 b0 b1 b2 b3 c0 c1 c2 c3 d0 d1 d2 d3 e0 e1 e2 e3 f0 f1 f2 f3 g0 g1 g2 g3 h0 h1 h2 h3 j0 j1 j2 j3 k0 k1 k2 k3 m0 m1 m2 m3 n0 n1 n2 n3 p0 p1 p2 p3 q0 q1 q2 q3 r0 r1 r2 r3 s0 s1 s2 s3 t0 t1 t2 t3 u0 u1 u2 u3 v0 v1 v2 v3 w0 w1 w2 w3 x0 x1 x2 x3 y0 y1 y2 y3 z0 z1 z2 z3
        )];
    $args{enc_world_length} ||= 2;

    $args{data} ||= [
        [
            [0,4,32,36],
            [8,12,40,44],
            [64,68,96,100],
            [72,76,104,108],
        ],
        [
            [1,5,33,37],
            [9,13,41,45],
            [65,69,97,101],
            [73,77,105,109],
        ],
        [
            [2,6,34,38],
            [10,14,42,46],
            [66,70,98,102],
            [74,78,106,110],
        ],
        [
            [3,7,35,39],
            [11,15,43,47],
            [67,71,99,103],
            [75,79,107,111],
        ],
        [
            [16,20,48,52],
            [24,28,56,60],
            [80,84,112,116],
            [88,92,120,124],
        ],
        [
            [17,21,49,53],
            [25,29,57,61],
            [81,85,113,117],
            [89,93,121,125],
        ],
        [
            [18,22,50,54],
            [26,30,58,62],
            [82,86,114,118],
            [90,94,122,126],
        ],
        [
            [19,23,51,55],
            [27,31,59,63],
            [83,87,115,119],
            [91,95,123,127],
        ],
    ];

    $class->SUPER::new(%args);
}

#my ($self, $lat, $lon, $timestamp, $level) = @_;
sub encode {
    my ($self, @pos) = @_;
    my $level = pop @pos;

    my $int = $self->_gen_int;
    my @enc = ();

    for my $i ( 1 .. $level ) {
        my $bits = 0;
        for my $j ( 0 .. 5 ) {
            my $flag = $j % 3;

            my $mid = ($int->[$flag][0] + $int->[$flag][1]) / 2;
            my $bit = $pos[$flag] >= $mid ? 1 : 0;

            $int->[$flag][$bit ? 0 : 1] = $mid;
            $bits = ( ( $bits << 1 ) | $bit );
        }
        my $mid = ($int->[2][0] + $int->[2][1])/2;
        my $bit = $pos[2] >= $mid ? 1 : 0;

        $int->[2][$bit ? 0 : 1] = $mid;
        $bits = ( ( $bits << 1 ) | $bit );
        push @enc, $self->{enc}[$bits];
    }

    join "", @enc;
}

sub decode_to_interval {
    my ($self, $str) = @_;

    my $int = $self->_gen_int;

    for my $ch ( $str =~ /.{$self->{enc_world_length}}/g ) {
        if ( defined ( my $bits = $self->{dec}{$ch} ) ) {
            for my $j ( 0 .. 5 ) {
                my $flag = $j % 3;
                $int->[$flag][ ( $bits & 64 ) >> 6 == 0 ? 1 : 0 ] = ( $int->[$flag][0] + $int->[$flag][1] ) / 2;
                $bits <<= 1;
            }
            $int->[2][ ( $bits & 64 ) >> 6 == 0 ? 1 : 0 ] = ( $int->[2][0] + $int->[2][1] ) / 2;
        }
        else {
            croak "Bad character '$ch' in hash '$str'";
        }
    }

    $int;
}

1;
__END__

=head1 NAME

Data::Wheren::7Bit - Perl extention to do something

=head1 VERSION

This document describes Data::Wheren::7Bit version 0.01.

=head1 SYNOPSIS

    use Data::Wheren::7Bit;

=head1 DESCRIPTION

# TODO

=head1 INTERFACE

=head2 Functions

=head3 C<< hello() >>

# TODO

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

<<YOUR NAME HERE>> E<lt><<YOUR EMAIL ADDRESS HERE>>E<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011, <<YOUR NAME HERE>>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
